File: Leetcode/1104.go
func f(level int, powerOfTwo []int) int {
	if level == 1 {
		return 1
	}
	return powerOfTwo[level-1] + powerOfTwo[level] - 1
}
func pathInZigZagTree(label int) []int {
	const MAXN int = 1e6 + 5
	var powerOfTwo []int
	powerOfTwo = append(powerOfTwo, 1)
	for ; powerOfTwo[len(powerOfTwo)-1] <= MAXN; {
		powerOfTwo = append(powerOfTwo, powerOfTwo[len(powerOfTwo)-1]<<1)
	}
	var ans []int
	var level = -1
	for i := 1; ; i++ {
		if powerOfTwo[i-1] <= label && label <= powerOfTwo[i]-1 {
			level = i
			break
		}
	}
	for {
		ans = append(ans, label)
		if label == 1 {
			break
		}
		if level&1 != 0 {
			label = f(level-1, powerOfTwo) - (label >> 1)
		} else {
			label = (f(level, powerOfTwo) - label) >> 1
		}
		level--
	}
	for i, j := 0, len(ans)-1; i < j; i, j = i+1, j-1 {
		ans[i], ans[j] = ans[j], ans[i]
	}
	return ans
}

File: Leetcode/1104/1104.cpp
constexpr int MAXN = 1e6+5;
class Solution {
public:
    vector<int> powerOfTwo;
    int getLevel(int num) {
        for (int i = 1;; ++i) {
            if (powerOfTwo[i-1] <= num && num <= powerOfTwo[i]-1) {
                return i;
            }
        }
        return -1;
    }
    int f(int level) {
        if (level == 1) {
            return 1;
        }
        return powerOfTwo[level-1]+powerOfTwo[level]-1;
    }
    vector<int> pathInZigZagTree(int label) {
        // pre
        vector<int>().swap(powerOfTwo);
        powerOfTwo.emplace_back(1);
        while (powerOfTwo.back() <= MAXN) {
            powerOfTwo.emplace_back(powerOfTwo.back()<<1);
        }
        vector<int> ans;
        int level;
        level = getLevel(label);
        while (true) {
            ans.emplace_back(label);
            if (label == 1) {
                break;
            }
            if (level & 1) {
                label = f(level-1)-(label>>1);
            }
            else {
                label = (f(level)-label)>>1;
            }
            --level;
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};